docs/CONFIG_DOC.md

Before getting started on config Make sure you have [Postgres](https://www.postgresql.org/) and [Postgis](https://postgis.net/) setup in your machine.


## Compulsary Configuration

### 1. Create ```config.txt``` inside src directory.
![image](https://user-images.githubusercontent.com/36752999/188402566-80dc9633-5d4e-479c-97dc-9e8a4999b385.png)


### 2. Prepare your OSM Snapshot Data
Initialize rawdata from [here](https://github.com/hotosm/underpass/tree/master/raw) OR Create database "raw" in your local postgres and insert sample dump of Pokhara city from
```
/tests/fixtures/pokhara.sql
```

```
psql -U postgres -h localhost raw < pokhara.sql
```
Put your credentials on Rawdata block

```
[RAW_DATA]
host=localhost
user=postgres
password=admin
database=raw
port=5432
```

### 3. Setup Oauth for Authentication
Login to [OSM](https://www.openstreetmap.org/) , Click on My Settings and register your local export_tool_api app to Oauth2applications

![image](https://user-images.githubusercontent.com/36752999/188452619-aababf28-b685-4141-b381-9c25d0367b57.png)


Check on read user preferences and Enter redirect URI as following
```
http://127.0.0.1:8000/latest/auth/callback/
```

Grab Client ID and Client Secret and put it inside config.txt as OAUTH Block , you can generate secret key for your application by yourself

```
[OAUTH]
client_id= your client id
client_secret= your client secret
url=https://www.openstreetmap.org
scope=read_prefs
login_redirect_uri=http://127.0.0.1:8000/latest/auth/callback/
secret_key=jnfdsjkfndsjkfnsdkjfnskfn
```

### 4. Configure celery and redis

API uses [Celery 5](https://docs.celeryq.dev/en/stable/getting-started/first-steps-with-celery.html) and [Redis 6](https://redis.io/download/#redis-stack-downloads) for task queue management , Currently implemented for Rawdata endpoint. 6379 is the default port . if you are running redis on same machine your broker could be ```redis://localhost:6379/```. You can change the port according to your configuration for the current docker compose use following

```
[CELERY]
CELERY_BROKER_URL=redis://redis:6379/0
CELERY_RESULT_BACKEND=redis://redis:6379/0
```

### 5. Finalizing config.txt
Insert your config blocks with the database credentials where you have underpass ,insight and rawdata in your database along with oauth block

Summary of command :

Considering You have PSQL-POSTGIS setup  with user **postgres** host **localhost** on port **5432** as password **admin**

```
  export PGPASSWORD='admin';
  psql -U postgres -h localhost -p 5432 -c "CREATE DATABASE raw;"

  cd tests/fixtures/
  psql -U postgres -h localhost -p 5432 raw  < pokhara.sql
```

Your config.txt will look like this

```
[RAW_DATA]
host=localhost
user=postgres
password=admin
database=raw
port=5432

[OAUTH]
client_id= your client id
client_secret= your client secret
url=https://www.openstreetmap.org
scope=read_prefs
login_redirect_uri=http://127.0.0.1:8000/latest/auth/callback/
secret_key=jnfdsjkfndsjkfnsdkjfnskfn

[API_CONFIG]
env=dev
log_level=debug
limiter_storage_uri=redis://redis:6379

[CELERY]
CELERY_BROKER_URL=redis://redis:6379/0
CELERY_RESULT_BACKEND=redis://redis:6379/0

```

**Tips** : Follow .github/workflows/[unit-test](https://github.com/hotosm/export-tool-api/blob/feature/celery/.github/workflows/unit-test.yml) If you have any confusion on implementation of config file .

## Optional Configuration [ You can skip this part for basic installation ]

You can further customize API if you wish with API_CONFIG Block

```
[API_CONFIG]
export_path=exports # used to store export path
api_host=http://127.0.0.1 # you can define this if you have different host
api_port=8000
max_area=100000 # max area to support for rawdata input
use_connection_pooling=True # default it will not use connection pooling but you can configure api to use to for psycopg2 connections
log_level=info #options are info,debug,warning,error
env=dev # default is dev , supported values are dev and prod
allow_bind_zip_filter=true # option to configure export output zipped/unzipped Default all output will be zipped
limiter_storage_uri=redis://localhost:6379 # API uses redis as backend for rate limiting
grid_index_threshold=5000 # value in sqkm to apply grid index filter
export_rate_limit=5 # no of requests per minute - default is 5 requests per minute
```
Based on your requirement you can also customize rawdata exports parameter using EXPORT_UPLOAD block

```
[EXPORT_UPLOAD]
FILE_UPLOAD_METHOD=disk # options are s3,disk , default disk
AWS_ACCESS_KEY_ID= your id
AWS_SECRET_ACCESS_KEY= yourkey
BUCKET_NAME= your bucket name
```



src/config.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# Copyright (C) 2021 Humanitarian OpenStreetmap Team

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU Affero General Public License as
# published by the Free Software Foundation, either version 3 of the
# License, or (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# You should have received a copy of the GNU Affero General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

# Humanitarian OpenStreetmap Team
# 1100 13th Street NW Suite 800 Washington, D.C. 20005
# <info@hotosm.org>

import errno
import logging
import os
from configparser import ConfigParser

from slowapi import Limiter
from slowapi.util import get_remote_address

CONFIG_FILE_PATH = "config.txt"
use_s3_to_upload = False

if os.path.exists(CONFIG_FILE_PATH) is False:
    logging.error(
        FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), CONFIG_FILE_PATH)
    )

config = ConfigParser()
config.read(CONFIG_FILE_PATH)

limiter_storage_uri = config.get(
    "API_CONFIG", "limiter_storage_uri", fallback="redis://localhost:6379"
)
# rate limiter for API requests based on the remote ip address and redis as backend
limiter = Limiter(key_func=get_remote_address, storage_uri=limiter_storage_uri)

export_rate_limit = int(config.get("API_CONFIG", "export_rate_limit", fallback=5))

grid_index_threshold = int(
    config.get("API_CONFIG", "grid_index_threshold", fallback=5000)
)

# get log level from config
log_level = config.get("API_CONFIG", "log_level", fallback=None)

if log_level is None or log_level.lower() == "debug":  # default debug
    level = logging.DEBUG
elif log_level.lower() == "info":
    level = logging.INFO
elif log_level.lower() == "error":
    level = logging.ERROR
elif log_level.lower() == "warning":
    level = logging.WARNING
else:
    logging.error(
        "logging config is not supported , Supported fields are : debug,error,warning,info , Logging to default :debug"
    )
    level = logging.DEBUG

# logging.getLogger("fiona").propagate = False  # disable fiona logging
logging.basicConfig(format="%(asctime)s - %(message)s", level=level)
logging.getLogger("boto3").propagate = False  # disable boto3 logging
logging.getLogger("botocore").propagate = False  # disable boto3 logging
logging.getLogger("s3transfer").propagate = False  # disable boto3 logging
logging.getLogger("boto").propagate = False  # disable boto3 logging


logger = logging.getLogger("export_tool_api")

export_path = config.get("API_CONFIG", "export_path", fallback=None)
if export_path is None:
    export_path = "exports"
if not os.path.exists(export_path):
    # Create a exports directory because it does not exist
    os.makedirs(export_path)
allow_bind_zip_filter = config.get("API_CONFIG", "allow_bind_zip_filter", fallback=None)
if allow_bind_zip_filter:
    allow_bind_zip_filter = True if allow_bind_zip_filter.lower() == "true" else False

AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY, BUCKET_NAME = None, None, None
# check either to use connection pooling or not
use_connection_pooling = config.getboolean(
    "API_CONFIG", "use_connection_pooling", fallback=False
)

# check either to use s3 raw data exports file uploading or not
file_upload_method = config.get(
    "EXPORT_UPLOAD", "FILE_UPLOAD_METHOD", fallback="disk"
).lower()
if file_upload_method == "s3":
    use_s3_to_upload = True
    try:
        AWS_ACCESS_KEY_ID = config.get("EXPORT_UPLOAD", "AWS_ACCESS_KEY_ID")
        AWS_SECRET_ACCESS_KEY = config.get("EXPORT_UPLOAD", "AWS_SECRET_ACCESS_KEY")
    except Exception as ex:
        logging.debug(ex)
        logging.debug("No aws credentials supplied")
    BUCKET_NAME = config.get(
        "EXPORT_UPLOAD", "BUCKET_NAME", fallback="exports-stage.hotosm.org"
    )
elif file_upload_method not in ["s3", "disk"]:
    logging.error(
        "value not supported for file_upload_method ,switching to default disk method"
    )
    use_s3_to_upload = False


def get_db_connection_params(dbIdentifier: str) -> dict:
    """Return a python dict that can be passed to psycopg2 connections
    to authenticate to Postgres Databases

    Params: dbIdentifier: Section name of the INI config file containing
            database connection parameters

    Returns: connection_params (dict): PostgreSQL connection parameters
             corresponding to the configuration section.

    """

    ALLOWED_SECTION_NAMES = ("UNDERPASS", "RAW_DATA")

    if dbIdentifier not in ALLOWED_SECTION_NAMES:
        logging.error(f"Invalid dbIdentifier. Pick one of {ALLOWED_SECTION_NAMES}")
        return None
    try:
        connection_params = dict(config.items(dbIdentifier))
        return connection_params
    except Exception as ex:
        logging.error(f"""Can't find DB credentials on config :{dbIdentifier}""")
        logging.error(ex)
